csrc/optimization/consecutive_cast.cpp

```diff
@@ -0,0 +1,194 @@
+  // TODO: Traveral implies topological order on returned exprs, we can leverage
+  // that to improve the effieciency of the pass. In the case of a straight line
+  // casts, we are doing a lot of meaningless work here on mutating intermediate
+  // casts that would have been done again at the end of the chain.
+  // We should really use the reverse topological order and filters out exprs
+  // that has been rendered as dead code during the pass.
+  for (auto expr : fusion->exprs()) {
+    // skip current expr if it's not a foldable cast
+    if (!isCast(expr)) {
+      continue;
+    }
+    std::list<Val*> chain_cast_vals;
+    auto prev_expr = expr->input(0)->definition();
+    while (isCast(prev_expr)) {
+      auto intermediate_cast = prev_expr->output(0);
+      // 1.2 Note, if the output of prev_expr
+      //   is used by other operation(s); or
+      //   is a direct output from fusion
+      // we skip the casting chaining
+      if (intermediate_cast->isFusionOutput() ||
+          intermediate_cast->uses().size() > 1) {
+        break;
+      }
+
+      // in the loop, we just repetitively chaining consecutive casts.
+      chain_cast_vals.push_front(intermediate_cast);
+      prev_expr = prev_expr->input(0)->definition();
+    }
+
+    // skip current expr if there's no chain_cast_vals
+    if (chain_cast_vals.empty()) {
+      continue;
+    }
+
+    // 1.3.1 Note, chain_cast_vals has a straight-line use without branches
+    auto lo_anchor = chain_cast_vals.front()->definition()->input(0);
+    auto anchor_dtype = lo_anchor->getDataType().value();
+    auto starting_anchor = lo_anchor;
+    for (auto val : chain_cast_vals) {
+      auto val_dtype = val->getDataType().value();
+
+      // 1.3.2.a short-cut when we are not losing precision
+      if (isInclusiveType(anchor_dtype, val_dtype)) {
+        continue;
+      }
+
+      // 1.3.2.c NOTE: To enter here, we have
+      //   !isInclusiveType(anchor_dtype, val_dtype) &&
+      //   !isInclusiveType(val_dtype, anchor_dtype)
+      //
+      // Which means the dtype between lo_anchor and val isn't compatible and
+      // can't be fold away without losing information. So we update the
+      // starting_anchor to current val, which ensures that we preserve the
+      // incompatible casts. e.g. for cases where no one type is strictly wider
+      // than the other: i.e. bf16 & fp16, int32 & float32 e.t.c.
+      if (!isInclusiveType(val_dtype, anchor_dtype)) {
+        lo_anchor = replaceInputInCast(lo_anchor, starting_anchor);
+        val = replaceInputInCast(val, lo_anchor);
+        // We need to update the starting_anchor for the fold to be past this
+        // current cast.
+        starting_anchor = val;
+      }
+      // 1.3.2.b/c updating new lo_anchor to current val
+      lo_anchor = val;
+      anchor_dtype = lo_anchor->getDataType().value();
+    }
+
+    auto output_dtype = expr->output(0)->getDataType().value();
+    if (anchor_dtype == output_dtype) {
```

Comment:
Could you add a test for this case? I don't think the logic inside this if clause is correct. For example, I didn't see the casts between start and anchor being removed.